File: pyscript/jvUtil/SQL.py
#!python3
# -*- coding:utf-8 -*-
import time
import requests


# 查询相关封装
class Construct:
    __acc = ""
    __token = ""
    __server_req_url = "http://jvquant.com/query/server?market=ab&type=sql&token="
    __sql_ser_addr = ""

    def __log(self, *args):
        print(time.strftime('%H:%M:%S', time.localtime(time.time())), args)

    def __init__(self, token, server="", logHandle=""):
        self.__token = token
        # 指定日志处理方法
        if logHandle != "":
            self.__log = logHandle
        # 如未指定server,则发起server查询
        if server != "":
            self.__server = server
        else:
            self.__getSerAddr()

    def __getSerAddr(self):
        url = self.__server_req_url + self.__token
        try:
            res = requests.get(url=url)
        except Exception as e:
            self.__log(e)
            return
        if (res.json()["code"] == "0"):
            self.__server = res.json()["server"]
            print("获取SQL服务器地址成功:" + self.__server)
        else:
            msg = "获取SQL服务器地址失败:" + res.text
            self.__log(msg)
            exit(-1)

    def qeury(self, query, page=1):
        url = "%s/sql?&mode=sql&acc=%s&token=%s&query=%s&page=%s" % (
            self.__server, self.__acc, self.__token, query, page)
        try:
            res = requests.get(url=url)
        except Exception as e:
            self.__log(e)
            return
        return res.json()

    def kline(self, code, cate="stock", type="day", fq="前复权", limit="240"):
        url = "%s/sql?&mode=kline&acc=%s&token=%s&cate=%s&code=%s&type=%s&fq=%s&limit=%s" % (
            self.__server, self.__acc, self.__token, cate, code, type, fq, limit)
        try:
            res = requests.get(url=url)
        except Exception as e:
            self.__log(e)
            return
        return res.json()

    def bond(self):
        url = "%s/sql?&mode=bond&acc=%s&token=%s" % (
            self.__server, self.__acc, self.__token)
        try:
            res = requests.get(url=url)
        except Exception as e:
            self.__log(e)
            return
        return res.json()

    def industry(self):
        url = "%s/sql?&mode=industry&acc=%s&token=%s" % (
            self.__server, self.__acc, self.__token)
        try:
            res = requests.get(url=url)
        except Exception as e:
            self.__log(e)
            return
        return res.json()

File: src/jvUtil/sql/sql.go
package sql

import (
	"bufio"
	"bytes"
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"time"
)

type ServerAddrRsp struct {
	Code   string `json:"code"`
	Server string `json:"server"`
}

type Sql struct {
	token  string //jvQuant Token
	server string
}

//初始化
func (sql *Sql) Construct(token, server string) {
	sql.token = token
	sql.server = server
	if server == "" {
		sql.server = sql.initServer()
	}
}

//获取数据库地址
func (sql *Sql) initServer() (server string) {
	params := url.Values{
		"market": []string{"ab"},
		"type":   []string{"sql"},
		"token":  []string{sql.token},
	}
	req := "http://jvquant.com/query/server?" + params.Encode()
	rb, err := HttpOnce(req, nil, nil, 3000)
	if err != nil {
		log.Fatalln("获取SQL服务器地址失败:", req, err)
	}
	rspMap := ServerAddrRsp{}
	err = json.Unmarshal(rb, &rspMap)
	if err != nil {
		log.Fatalln("解析SQL服务器地址异常:", string(rb), err)
	}
	server = rspMap.Server
	if rspMap.Code != "0" || server == "" {
		log.Fatalln("获取SQL服务器地址异常:", string(rb))
	}
	log.Println("获取SQL服务器地址成功:", server)
	return
}

//语义泛查
func (sql *Sql) Query(query, page string) (rb []byte) {
	params := url.Values{
		"token": []string{sql.token},
		"mode":  []string{"sql"},
		"query": []string{query},
		"page":  []string{page},
	}
	req := sql.server + "/sql?" + params.Encode()
	rb, err := HttpOnce(req, nil, nil, 1000)
	if err != nil {
		log.Println("SQL请求泛查异常:", req, err)
	}
	return
}

//获取K线
func (sql *Sql) Kline(code, cate, _type, fq, limit string) (rb []byte) {
	params := url.Values{
		"token": []string{sql.token},
		"mode":  []string{"kline"},
		"cate":  []string{cate},
		"code":  []string{code},
		"type":  []string{_type},
		"fq":    []string{fq},
		"limit": []string{limit},
	}
	req := sql.server + "/sql?" + params.Encode()
	rb, err := HttpOnce(req, nil, nil, 1000)
	if err != nil {
		log.Println("SQL请求K线异常:", req, err)
	}
	return
}

//获取可转债信息
func (sql *Sql) Bond() (rb []byte) {
	params := url.Values{
		"token": []string{sql.token},
		"mode":  []string{"bond"},
	}
	req := sql.server + "/sql?" + params.Encode()
	rb, err := HttpOnce(req, nil, nil, 1000)
	if err != nil {
		log.Println("SQL请求转债信息异常:", req, err)
	}
	return
}

//获取行业分类信息
func (sql *Sql) Industry() (rb []byte) {
	params := url.Values{
		"token": []string{sql.token},
		"mode":  []string{"industry"},
	}
	req := sql.server + "/sql?" + params.Encode()
	rb, err := HttpOnce(req, nil, nil, 1000)
	if err != nil {
		log.Println("SQL请求行业分类异常:", req, err)
	}
	return
}

func HttpOnce(Url string, headers, postData map[string]string, msTimeOut int) (r []byte, err error) {
	client := &http.Client{
		Timeout: time.Duration(time.Duration(msTimeOut) * time.Millisecond),
	}
	method := http.MethodGet
	r = []byte{}
	err = nil
	if len(headers) == 0 {
		headers = map[string]string{}
	}
	if len(postData) != 0 {
		method = http.MethodPost
		headers["Content-Type"] = "application/x-www-form-urlencoded"
	}

	postParam := url.Values{}
	for k, v := range postData {
		postParam.Set(k, v)
	}
	postParamBuff := bytes.NewBufferString(postParam.Encode())
	req, err := http.NewRequest(method, Url, postParamBuff)

	if err != nil {
		return r, err
	}
	for k, v := range headers {
		req.Header.Add(k, v)
	}
	resp, er := client.Do(req)
	if er != nil {
		err = er
		return
	}
	defer resp.Body.Close()
	if err != nil {
		return r, err
	}
	br := bufio.NewReader(resp.Body)
	r, err = ioutil.ReadAll(br)
	return r, err
}
